aws-auth-cognito/src/test/java/com/amplifyframework/auth/cognito/AuthComponentTest.java

```diff
@@ -0,0 +1,459 @@
+            return null;
+        }).when(mobileClient).initialize(any(Context.class), any(AWSConfiguration.class), any(Callback.class));
+
+        authPlugin.configure(testConfig, context);
+        verify(mobileClient, times(1))
+                .initialize(any(Context.class), any(AWSConfiguration.class), any(Callback.class));
+    }
+
+    /**
+     * Tests that if AWSMobileClient returns an error callback during initialization, the Auth configure method
+     * throws an AuthException.
+     * @throws AuthException the AuthException expected to be thrown if initialization fails.
+     */
+    @Test(expected = AuthException.class)
+    public void testConfigureExceptionHandling() throws AuthException {
+        doAnswer(invocation -> {
+            Callback<UserStateDetails> callback = invocation.getArgument(2);
+            callback.onError(new Exception());
+            return null;
+        }).when(mobileClient).initialize(any(Context.class), any(AWSConfiguration.class), any(Callback.class));
+
+        authPlugin.configure(new JSONObject(), getApplicationContext());
+    }
+
+    /**
+     * Tests that the signUp method of the Auth wrapper of AWSMobileClient (AMC) calls AMC.signUp() with the username
+     * and password it received and, if included, the userAttributes and validationData sent in the options object.
+     * Also ensures that in the onResult case, the success callback receives a valid AuthSignUpResult.
+     */
+    @Test
+    public void signUp() {
+        SignUpResult amcResult = new SignUpResult(
+            false,
+            new UserCodeDeliveryDetails(
+                    destination,
+                    deliveryMedium,
+                    attributeName
+            ),
+            null
+        );
+
+        doAnswer(invocation -> {
+            assertEquals(username, invocation.getArgument(0));
+            assertEquals(password, invocation.getArgument(1));
+            Map<String, String> attributeMap = invocation.getArgument(2);
+            assertTrue(attributeMap.containsKey(attributeKey));
+            assertEquals(attributeVal, attributeMap.get(attributeKey));
+            assertEquals(metadata, invocation.getArgument(3));
+
+            Callback<SignUpResult> callback = invocation.getArgument(4);
+            callback.onResult(amcResult);
+            return null;
+        }).when(mobileClient)
+            .signUp(any(String.class), any(String.class), any(Map.class), any(Map.class), any(Callback.class));
+
+        authPlugin.signUp(
+            username,
+            password,
+            AWSCognitoAuthSignUpOptions.builder()
+            .userAttributes(
+                Collections.singletonList(new AuthUserAttribute(AuthUserAttributeKey.email(), attributeVal))
+            )
+            .validationData(metadata)
+            .build(),
+            result -> validateSignUpResult(result, AuthSignUpStep.CONFIRM_SIGN_UP_STEP),
+            error -> fail());
```

Comment:
If this callback occurs on a different thread, `fail()` may not do what you think it will. `fail()` needs to be called on the _test runner_ thread, to halt the test. If a background thread crashes, but the `@Test` has already moved on, jUnit won't catch it.

TLDR: use something like `Await.result` to pull the exception back into your calling thread, and then do something with it, from that thread. `Completable.fromAction(() -> {}).blockingAwait()` is another way to achieve the same.

Reply:
Ok, I may need to sync up with you on this